pkg/adapter/adapter.go

```diff
@@ -31,17 +35,150 @@ type envConfig struct {
+	}
+	URI := string(rawURI)
+	if URI == "" {
+		logger.Desugar().Error("MongoDb URI field is an empty string", zap.Any("URI", URI))
+	}
+
+	return newAdapter(ctx, env, ceClient, URI)
+}
+
+func newAdapter(ctx context.Context, env *envConfig, ceClient cloudevents.Client, URI string) adapter.Adapter {
+	logger := logging.FromContext(ctx)
+
+	// Create new Client.
+	client, err := mongo.NewClient(options.Client().ApplyURI(URI))
+	if err != nil {
+		logger.Desugar().Error("Error creating mongo client", zap.Error(err))
+	}
+	// Get dataSource
+	dataSource := getDataSource(client, env.Database, env.Collection)
+
+	return &mongoDbAdapter{
+		namespace:  env.Namespace,
+		ce:         ceClient,
+		client:     client,
+		datasource: dataSource,
+		source:     env.EventSource,
+		sinkurl:    env.SinkURL,
+		logger:     logger,
+	}
+}
+
+// Returns the appropriate dataSource: database or collection.
+func getDataSource(client *mongo.Client, db string, coll string) *dataSource {
+	if coll != "" {
+		var collsource dataSource = client.Database(db).Collection(coll)
+		return &collsource
+	}
+	var dbsource dataSource = client.Database(db)
+	return &dbsource
+}
+
+// Start creates the watch stream that will watch for dataSource changes.
+func (a *mongoDbAdapter) Start(ctx context.Context) error {
+	// Connect to Client.
+	err := a.client.Connect(ctx)
+	if err != nil {
+		a.logger.Desugar().Error("Error connecting to database", zap.Error(err))
+		return err
+	}
+	defer a.client.Disconnect(ctx)
+
+	// Create a watch stream for either the database or collection.
+	var datasource dataSource = *a.datasource
+	stream, err := datasource.Watch(ctx, mongo.Pipeline{})
+
+	if err != nil {
+		a.logger.Desugar().Error("Error setting up changeStream", zap.Error(err))
+		return err
+	}
+	defer stream.Close(ctx)
+
+	// Watch and process changes.
+	err = a.processChanges(ctx, stream)
+	if err != nil {
+		a.logger.Desugar().Error("Error watching and processing changes", zap.Error(err))
+		return err
+	}
 	return nil
 }
+
+// processChanges processes the new incoming change, creates a cloud event and sends it.
+func (a *mongoDbAdapter) processChanges(ctx context.Context, stream *mongo.ChangeStream) error {
+	// For each new change recorded.
+	for stream.Next(ctx) {
+		var data bson.M
+		if err := stream.Decode(&data); err != nil {
+			a.logger.Desugar().Error("Error decoding the change stream", zap.Error(err))
+		}
+		// Create corresponding event.
+		event, err := makeCloudEvent(data)
+		if err != nil {
+			a.logger.Desugar().Error("Failed to create event", zap.Error(err))
+		}
+
+		// Send event using client.
+		ctx := cloudevents.ContextWithTarget(ctx, a.sinkurl)
+		a.logger.Desugar().Info("", zap.Any("sinkurl", a.sinkurl))
+
+		// Send that Event.
+		if result := a.ce.Send(ctx, *event); cloudevents.IsUndelivered(result) {
+			a.logger.Desugar().Error("Failed to send event", zap.Any("result", result))
+		}
```

Comment:
If we fail to send the event, do we lose the event for good? Will mongodb retry the same event later?

Reply:
It is lost for good I think, we can talk about alternatives during standup

Reply:
Look into how mongodb deals with resending